### ISSUE TITLE
On launch, tabs open and steal focus unexpectedly

### ISSUE THREAD

**Issue by salty-horse** [hexchat/hexchat#1676]
Using HexChat 2.10.2

I have several servers set to autoconnect on launch.

If the connection takes a long time, it may happen that I will be on one tab, already typing something in the text box, then, once another server connects, a tab for a different channel will open and steal the focus, and before noticing, I will end up typing an ending of a message and hitting ENTER, sending a truncated the message to the wrong channel.

I can think of several solutions:

1. tabs that are opening due to an autoconnect should not steal focus.
2. If the text box is not empty, new tabs shouldn't steal focus.

**Comment by TingPing** [comment]
1 is the solution to me, the option (*Preferences > Channel Switcher > Focus New Tabs*) even says as much though it doesn't seem to really do it. Do you happen to use ZNC?

**Comment by salty-horse** [comment]
I don't use ZNC or any other bouncer.

**Comment by badon** [comment]
I was not able to duplicate this bug. However, if I use the "connect commands" feature to join channels (it's easier to manage lots of channels that way), upon disconnection and auto-reconnect, it will switch tabs until all of the hundreds of channels are rejoined. This can make Hexchat unusable for several minutes, depending on how much delay there is on all of the join commands. Repeated disconnections on one bad server makes all of the channels unusable. I'm guessing salty-horse is using connect commands? That's the only way I have been able to get Hexchat to disobey the option in:

Preferences --> Channel Switcher --> Focus New Tabs

I tested with 2.12.1. My method of breaking Hexchat's obedience to the Focus New Tabs feature is a separate bug.

**Comment by TingPing** [comment]
@badon Yes `/join` ran when you already are in a channel will focus it manually. I suggest to just use the real autojoin feature...

**Comment by badon** [comment]
The thing is, since Hexchat was disconnected, it is NOT in ANY channels. It appears the focus feature of /join is unaware of when it is not really connected to a channel (even though the tab for the channel exists). I have noticed that sometimes a DUPLICATE tab will be opened, leaving behind a non-functioning channel tab with parentheses around the channel name. I suspect that is also another bug caused by /join's lack of awareness of whether Hexchat is actually in a channel, or just has a tab for the channel in a disconnected state.

I haven't been able to duplicate the "duplicate tab" bug, so I'm unsure if it's related to salty-horse's bug. I suspect it is related, since that's the only way I have been able to duplicate the behavior reported here.

**Comment by badon** [comment]
> If the connection takes a long time

That's the only clue that in salty-horse's report that indicates a disconnection may be involved in triggering his experience of the bug. If I'm correct in guessing that, then its' more likely that his bug is indeed the same one that I've documented here.

**Comment by TingPing** [comment]
> It appears the focus feature of /join is unaware of when it is not really connected to a channel

Indeed.

> I have noticed that sometimes a DUPLICATE tab will be opened

Seems possible, hexchat has a concept of a channel that will be rejoined on reconnect and reuses the tabs.

Anyway your usage is very different so if you think the `/join` behavior is doing something weird I suggest its own issue.

**Comment by salty-horse** [comment]
> That's the only clue in salty-horse's report that indicates a disconnection may be involved in triggering his experience of the bug.

No disconnects.

Here's my specific use-case:
* Several servers set to auto-connnect by having their "Connect to this network automatically" setting turned on.
* Those servers all have "autojoin channels" listed.
* The "Focus new tabs" setting is set to "Only requested tabs" - I haven't fiddled with it. It's the default.

When I launch HexChat, I get a tab for every server. They try and connect in parallel. Once each connects, it autojoins channels, each having their own window gain focus.

**Comment by badon** [comment]
@TingPing Posted about /join focus stealing after reconnecting disconnected channel tabs: https://github.com/hexchat/hexchat/issues/1719

**Comment by postmodern** [comment]
I also am annoyed by this behavior, although I use a server/channel list. While I wait for slower servers to login, I may click on other channels in the list. Once the slower servers connect and I auto-join my various channels, it populates the channel list, but my focus stays at the given index and doesn't stay with the channel that I originally selected.
